cbor2/decoder.py

```diff
@@ -43,15 +44,15 @@ def decode_bytestring(decoder, subtype, shareable_index=None):
     length = decode_uint(decoder, subtype, allow_indefinite=True)
     if length is None:
         # Indefinite length
-        buf = bytearray()
+        buf = []
         while True:
             initial_byte = byte_as_integer(decoder.read(1))
             if initial_byte == 255:
-                return buf
+                return b''.join(buf)
```

Comment:
I'm concerned about returning bytes instead of a bytearray here. Have you checked how much, if any, impact this has on memory use and performance?

Reply:
I'm sure buffering the string fragments in a list will increase the memory usage. I'm not sure how I could test it's impact.

I believe when using strings, the ''.join(<strings>) method is way more efficient than repeated concatenation. bytearrays probably do it differently though.

Reply:
I've updated the Gist, at least on python 3.6, calling join on a list of bytes is about 3 times faster and uses half the memory of appending to a bytearray.

Reply:
If the size of each string is 100 bytes and the number of strings to be concatenated is 500,000 you get:

```python
%%timeit -r 10 -n 100000 data = gen(100,500000)
concat_bytes(data)
```

    100000 loops, best of 10: 14.4 µs per loop
    


```python
%%timeit  -r 10 -n 100000 data = gen(100,500000)
concat_join(data)
```

    100000 loops, best of 10: 12.8 µs per loop
    


```python
%%memit data = gen(100,500000)
concat_bytes(data)
```

    peak memory: 97.41 MiB, increment: 44.68 MiB
    


```python
%%memit data = gen(100,500000)
concat_join(data)
```

    peak memory: 162.55 MiB, increment: 110.62 MiB

Reply:
Alright, you've convinced me.